templates/default.nginx_http.conf.template

```diff
@@ -21,6 +21,14 @@ server_tokens off;
 # Don't allow page to be rendered inside a frame or iframe
 add_header X-Frame-Options DENY;
 
+# Enable cross-site scripting filter (XSS)
+# (handled by the browser; usually enabled by default,
+# this header will re-enable the filter if disabled by user.)
+add_header X-XSS-Protection "1; mode=block";
+
+# Explicitly list domains allowed to serve content for this site
+add_header Content-Security-Policy "default-src https: 'self' 'unsafe-inline' 'unsafe-eval' https://gui.dockstore.org https://fonts.googleapis.com; img-src 'self' https://gui.dockstore.org https://*.gravatar.com; object-src 'none';";
```

Comment:
Does this policy also affect html content? We have featured content that comes from an s3 bucket.

Reply:
Same idea, does this affect pictures for organizations, Twitter widget, images from workflow markdown descriptions?

Reply:
That is correct - it affects iframes, media, images, fonts, css, javascript, icons, plugins, and a few other things.

I inspected several pages at dockstore.org to see where images were being hosted, and the large majority (dockstore icons, university logos, menu buttons, etc) were hosted at gui.dockstore.org URLs. I'm not sure if that's the bucket content you're referring to, but if so, it looks like the content is being transparently wrapped with the gui.dockstore.org URL.

A better way to do this (found after I had requested reviews on this PR) is to use report-uri.com's CSP generator (<https://report-uri.com/home/generate#tab_1_18>), which can load a URL, track all content that was loaded, and assemble a CSP that will explicitly allow all the content on that page. Running that on a few different dockstore.org pages should help cover anything my manually-written CSP missed.

Reply:
> @coverbeck I am not sure what content is coming from an S3 bucket ("content" refers to content embedded into the HTML, not HTML itself)

The Explore Featured Content section of the logged-in user's home page is coming from an s3 bucket: `https://s3.amazonaws.com/dockstore.featured.content/production/feat-content.html`. It is HTML that we are fetching and inserting into the DOM.

Which I guess brings up a point when using the CSP generator; make sure to go to the home page being both logged in and not logged in.

Reply:
So I'm thinking about stuff like the organisations page where icons are specified from anywhere https://dockstore.org/organizations/BroadInstitute 
The twitter widget is on the right side of the logged in home page
Badges aren't too common, but one example from quay.io is on the right side of https://dockstore.org/containers/quay.io/pancancer/pcawg-sanger-cgp-workflow:2.1.0?tab=info

Reply:
I think the org icons are covered by our usage of Gravatar, which we use to sanitize  images.

But I'm concerned about the other cases too; and what if the Twitter widget has an image that comes from yet another web site (or maybe they don't allow that)?

Reply:
> It is HTML that we are fetching and inserting into the DOM.

Thank you for clarifying. Initially I thought that this insertion operation would not be affected by the CSP header, but then I found a [`connect-src` directive](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/connect-src) for the CSP header that restricts domains that Javascript can connect to. If we use any of the following JS functions (assuming we are using one of these to grab the HTML file from the bucket), we will need to add the S3 bucket URL to a connect-src directive:

- <a> ping,
- Fetch,
- XMLHttpRequest,
- WebSocket,
- EventSource
- Navigator.sendBeacon()

Reply:
I think it's using XMLHttpRequest. We're using Angular HttpClient, which I'm pretty sure uses XMLHttpRequest under the hood. I had to look up Fetch; didn't know it existed. :) Same difference though, since they're both in the list.

Reply:
> what if the Twitter widget has an image that comes from yet another web site (or maybe they don't allow that

This should be covered by the CSP Generator tool that I used - if the Twitter widget did load images from another website, it would have shown up in the generated CSP.